Animate the user account drawer header arrow

dudeofawesome opened the pull request:
When the user account drawer header is tapped, the arrow in the bottom right will now rotate instead of just switching to the upward facing arrow.

dudeofawesome committed:
🚧 convert _AccountDetails to stateful widget

dudeofawesome committed:
🚧 convert _AccountDetails to stateful widget

dudeofawesome committed:
🚧 convert _AccountDetails to stateful widget

dudeofawesome committed:
✨💄 animate account switcher arrow

dudeofawesome committed:
✨💄 animate account switcher arrow

dudeofawesome committed:
✨💄 animate account switcher arrow

dudeofawesome committed:
✨💄 animate account switcher arrow

dudeofawesome committed:
🐛 set initial arrow position by isOpen

dudeofawesome committed:
🐛 set initial arrow position by isOpen

dudeofawesome committed:
🐛 set initial arrow position by isOpen

dudeofawesome committed:
🎨 prefer declare const constructors

dudeofawesome committed:
🎨 prefer declare const constructors

dudeofawesome committed:
🎨 prefer declare const constructors

Hixie commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -60,75 +62,119 @@ class _AccountPictures extends StatelessWidget {
+        curve: Curves.easeInOut,
+      )
+      ..addListener(() {
+        setState(() {});
```
always put a comment inside the block of a setState() if it's otherwise empty, explaining what it is that the build method depends on that has changed

Hixie commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -60,75 +62,119 @@ class _AccountPictures extends StatelessWidget {
+
+    _controller = new AnimationController(
+      value: widget.isOpen ? 1.0 : 0.0,
+      duration: const Duration(milliseconds: 200),
```
why 200ms?

dudeofawesome replied:
I picked 200 ms because that's the duration of the account switcher animations in the gallery's drawer example.

https://github.com/flutter/flutter/blob/master/examples/flutter_gallery/lib/demo/material/drawer_demo.dart#L36

Hixie commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -60,75 +62,119 @@ class _AccountPictures extends StatelessWidget {
+    );
+    _animation = new CurvedAnimation(
+        parent: _controller,
+        curve: Curves.easeInOut,
```
why this curve?

dudeofawesome replied:
I picked `easeInOut` because that's what [the Material guidelines recommend](https://material.io/guidelines/motion/duration-easing.html#duration-easing-natural-easing-curves). Should this be a different curve?

Hixie replied:
Ah, I see. That's actually `Curves.fastOutSlowIn` (the name matches Android's name for the same).

Hixie commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -60,75 +62,119 @@ class _AccountPictures extends StatelessWidget {
+
+  void _onTap() {
+    if (widget.isOpen) {
+      _controller.reverse();
```
this doesn't seem right. Consider the case where the user taps the widget, onTap is invoked, and does _not_ change isOpen. Now the arrow will be animating the wrong direction.

I think it would make more sense to trigger the animation during didUpdateWidget based on the new value of isOpen.

See `data_table.dart` in the material directory, there's a _SortArrow widget there that you may wish to crib from. (It also does other cool things, like always going clockwise unless you change it mid-turn, IIRC.)

Hixie commented:
this will need tests

dudeofawesome committed:
👌💬 add comment to setState call

dudeofawesome committed:
👌💬 add comment to setState call

dudeofawesome committed:
👌💬 add comment to setState call

dudeofawesome committed:
👌💄 use fastOutSlowIn curve

dudeofawesome committed:
👌💄 use fastOutSlowIn curve

dudeofawesome committed:
👌💄 use fastOutSlowIn curve

dudeofawesome commented:
@Hixie What would you recommend I test? I'm not very familiar with UI testing in Flutter.

dudeofawesome committed:
🎨 simplify animation listener structure

dudeofawesome committed:
🎨 simplify animation listener structure

dudeofawesome committed:
🎨 simplify animation listener structure

dudeofawesome committed:
👌🐛 fix animation playing backwards if onTap doesn't act as expected

dudeofawesome committed:
👌🐛 fix animation playing backwards if onTap doesn't act as expected

dudeofawesome committed:
👌🐛 fix animation playing backwards if onTap doesn't act as expected

dudeofawesome committed:
👌🐛 fix animation playing backwards if onTap doesn't act as expected

dudeofawesome committed:
🐛💄 use flipped curve for arrow return

dudeofawesome committed:
🐛💄 use flipped curve for arrow return

dudeofawesome committed:
🐛💄 use flipped curve for arrow return

Hixie commented:
@dudeofawesome Anything you don't want us to regress should be tested. Imagine that we have a bot which randomly changes a line of code, then runs the tests, and if the tests pass, it checks the change in.
So for example, you could test that if the user taps on the header, the arrow animates correctly.

Hixie commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -60,75 +62,121 @@ class _AccountPictures extends StatelessWidget {
+    super.didUpdateWidget(oldWidget);
+
+    if (widget.isOpen) {
+      _controller.forward();
```
This will cause the animation to repeatedly restart each time the widget is rebuilt, if the direction didn't change. You probably want to check if the animation is already running, and only start it if it's not running in the right direction. So e.g. in this case you'd want to call forward() if the animation status was dismissed or reverse, and where you call reverse you only want to call it if the status is completed or forward.

Hixie commented:
@dudeofawesome I'm going to close this PR since you're not waiting for review right now, but please don't hesitate to reopen it, or submit a new one, once you are ready for further review. Thanks!

If you don't want to work further on this PR that's fine too, I've tagged it so that when we have some free time we can fix whatever remains to be fixed and land it for you.

dudeofawesome committed:
🔀 Merge remote-tracking branch 'flutter/master'

dudeofawesome committed:
🔀 Merge remote-tracking branch 'upstream/master'

dudeofawesome commented:
@Hixie I've been trying to get this updated with the latest version of Flutter, but have been stuck on a an odd pub issue. It won't `get` the gallery demo's packages because it says it can't find version `0.3.0` of the `connectivity` package. 

```
$ flutter packages get
Running "flutter packages get" in flutter_gallery...
Package connectivity has no versions that match 0.3.0 derived from:
- flutter_gallery depends on version 0.3.0
pub get failed (1)
```

I've upgraded my system Dart install to the latest (`2.0.0-dev.43.0 (Wed Mar 28 01:04:15 2018 +0200) on "macos_x64"`), and am using the bundled `flutter` command:
```
Flutter 0.0.21-pre.336 • channel feat/user-account-arrow-anim • https://github.com/dudeofawesome/flutter.git
Framework • revision 53810b82d1 (50 minutes ago) • 2018-04-01 17:43:56 -0700
Engine • revision 6280adbfb1
Tools • Dart 2.0.0-dev.39.0.flutter-06949dc985
```
I've merged in the latest changes from `upstream/master`.

`flutter run` attempts to upgrade before it does anything else, so I can't really do anything until I get this fixed. Do you have any suggestions on how I might fix that?

Hixie commented:
Hm, I wonder if what's going on is that your local changes are confusing our version code, and so pub is refusing to give you packages that relate to more recent changes.

Can you try `git fetch`, to make sure you have the new git tags downloaded, and if that doesn't help, try rebasing so that your branch is up to date with your commits on top rather than having a separate branch with a very old merge point? (I don't know if what I just said makes sense, git mystifies me a bit.)

dudeofawesome committed:
Merge remote-tracking branch 'upstream/master'

dudeofawesome commented:
I just tried `git fetch --all` and `git rebase master` to get latest tags and such, which doesn't seem to have helped. I also tried switching branch to master and running `flutter update-packages`, which ran in to the same error.

Is it possible that pub's got some broken package manifest cached?

jonahwilliams commented:
@dudeofawesome I've seen this error multiple times today, the root cause was always a misconfigured flutter repo and not pub. Your flutter version should be something like:
```
Flutter 0.2.8 • channel unknown • unknown source
Framework • revision b397406561 (5 hours ago) • 2018-04-02 13:53:20 -0700
Engine • revision c903c217a1
Tools • Dart 2.0.0-dev.43.0.flutter-52afcba357
```
CC @DaveShuckerow who also had and fixed this problem today.

dudeofawesome commented:
For this project, I'm using the bundled flutter in bin on the master branch, which shows that I have `0.0.21-pre.328`
```
$ git branch
  feat/user-account-arrow-anim
* master
$ which flutter
/Users/dudeofawesome/Github/flutter/bin/flutter
$ bin/flutter --version
Flutter 0.0.21-pre.328 • channel master • https://github.com/dudeofawesome/flutter.git
Framework • revision 761eb80fd5 (35 minutes ago) • 2018-04-02 18:25:49 -0700
Engine • revision c903c217a1
Tools • Dart 2.0.0-dev.43.0.flutter-52afcba357
```

Should I be using my system `flutter` instead? It's currently on the beta branch with version `0.2.8`

dudeofawesome commented:
I just tried using my system `flutter` instead of the bundled one, and despite the warning, it worked fine! 

The [`CONTRIBUTING.md`](https://github.com/flutter/flutter/blob/master/CONTRIBUTING.md) on master doesn't mention anything about a preferred branch to base from. Is `master` the correct one to use right now, or is `beta` the correct choice?

dudeofawesome committed:
Merge branch 'feat/user-account-arrow-anim' of https://github.com/dudeofawesome/flutter into feat/user-account-arrow-anim

Hixie commented:
can you elaborate on what you mean by "bundled flutter"?

for development, use master.

dudeofawesome commented:
The "bundled" `flutter` version I'm talking about is in my branch's bin folder [`/bin/flutter`](https://github.com/dudeofawesome/flutter/blob/feat/user-account-arrow-anim/bin/flutter).

dudeofawesome committed:
Merge remote-tracking branch 'upstream/master'

dudeofawesome committed:
🔀 Merge remote-tracking branch 'upstream/master'

dudeofawesome committed:
🔀 Merge branch 'master' into feat/user-account-arrow-anim

engine-flutter-autoroll committed:
Roll engine 3236b49ceac5..162b2e98c036 (1 commits) (#23593)


https://github.com/flutter/engine/compare/3236b49ceac5...162b2e98c036

git log 3236b49ceac5..162b2e98c036 --no-merges --oneline
162b2e98c Roll src/third_party/skia a944142412cd..3b79aa3a5ad0 (15 commits) (flutter/engine#6671)


The AutoRoll server is located here: https://autoroll.skia.org/r/flutter-engine-flutter-autoroll

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/&#43;/master/autoroll/README.md

If the roll is causing failures, please contact the current sheriff, who should
be CC&#39;d on the roll, and stop the roller if necessary.

engine-flutter-autoroll committed:
Roll engine 162b2e98c036..6c2a0b3a6d93 (5 commits) (#23612)


https://github.com/flutter/engine/compare/162b2e98c036...6c2a0b3a6d93

git log 162b2e98c036..6c2a0b3a6d93 --no-merges --oneline
6c2a0b3a6 Undefine ERROR in platform_view_layer.cc (flutter/engine#6675)
55e12993a Update FlutterPlugin.h docs, suppress warning for older API (flutter/engine#6672)
f7970048d Attach and position embedded UIVIews (flutter/engine#6614)
df85722fa Plumb the iOS PlatformViewsController into flow. (flutter/engine#6603)
2bfb893cf iOS Embedding Refactor (flutter/engine#6447)


The AutoRoll server is located here: https://autoroll.skia.org/r/flutter-engine-flutter-autoroll

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/&#43;/master/autoroll/README.md

If the roll is causing failures, please contact the current sheriff, who should
be CC&#39;d on the roll, and stop the roller if necessary.

liyuqian committed:
Revert "Roll engine 162b2e98c036..6c2a0b3a6d93 (5 commits) (#23612)" (#23628)

This reverts commit 0f34f526204d180e117d2fc296084155651cf8da.

Reverts flutter/flutter#23612 as it broke the following two bots:
1. platform_view__start_up
2. platform_interaction_test_ios

Merge on red to fix the rope.

Hixie committed:
Fix a race condition in vmservice_test.dart (#23529)

Fixes https://github.com/flutter/flutter/issues/19273

Hixie committed:
[H] Created a variant of InheritedWidget specifically for Listenables (#23393)

* Created a variant of InheritedWidget specifically for Listenables

* Add more documentation per review comments

Hixie committed:
Revert "[O] Fix a race condition in vmservice_test.dart" (#23647)

* Revert "[H] Created a variant of InheritedWidget specifically for Listenables (#23393)"

This reverts commit 931328596ac48a848953f330a192fc33425a378a.

* Revert "Fix a race condition in vmservice_test.dart (#23529)"

This reverts commit 5e7b0a366b3fe55b99fa79461a94140c59ab3f07.

engine-flutter-autoroll committed:
Roll engine 162b2e98c036..a1d7cad70c81 (10 commits) (#23646)


https://github.com/flutter/engine/compare/162b2e98c036...a1d7cad70c81

git log 162b2e98c036..a1d7cad70c81 --no-merges --oneline
a1d7cad70 Fix inconsistent include syntax (flutter/engine#6680)
236661cd8 Roll src/third_party/skia 3b79aa3a5ad0..b53f1f46982d (13 commits) (flutter/engine#6679)
c4f50611d Roll buildroot to pick up Mac toolchain updates. (flutter/engine#6678)
ba8f6aa71 Handle Windows headers defining ERROR to 0 in log levels. (flutter/engine#6677)
505d2a923 Roll buildroot to pick up updates to custom toolchains. (flutter/engine#6674)
6c2a0b3a6 Undefine ERROR in platform_view_layer.cc (flutter/engine#6675)
55e12993a Update FlutterPlugin.h docs, suppress warning for older API (flutter/engine#6672)
f7970048d Attach and position embedded UIVIews (flutter/engine#6614)
df85722fa Plumb the iOS PlatformViewsController into flow. (flutter/engine#6603)
2bfb893cf iOS Embedding Refactor (flutter/engine#6447)


The AutoRoll server is located here: https://autoroll.skia.org/r/flutter-engine-flutter-autoroll

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/&#43;/master/autoroll/README.md

If the roll is causing failures, please contact the current sheriff, who should
be CC&#39;d on the roll, and stop the roller if necessary.

engine-flutter-autoroll committed:
Roll engine a1d7cad70c81..2f06a53a8d19 (1 commits) (#23654)


https://github.com/flutter/engine/compare/a1d7cad70c81...2f06a53a8d19

git log a1d7cad70c81..2f06a53a8d19 --no-merges --oneline
2f06a53a8 Roll src/third_party/skia b53f1f46982d..38e4fd0c5654 (1 commits) (flutter/engine#6681)


The AutoRoll server is located here: https://autoroll.skia.org/r/flutter-engine-flutter-autoroll

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/&#43;/master/autoroll/README.md

If the roll is causing failures, please contact the current sheriff, who should
be CC&#39;d on the roll, and stop the roller if necessary.

engine-flutter-autoroll committed:
Roll engine 2f06a53a8d19..7352251ebec6 (2 commits) (#23667)


https://github.com/flutter/engine/compare/2f06a53a8d19...7352251ebec6

git log 2f06a53a8d19..7352251ebec6 --no-merges --oneline
7352251eb Roll src/third_party/skia 79c96811863f..797197a772b8 (1 commits) (flutter/engine#6683)
3dac47e4d Roll src/third_party/skia 38e4fd0c5654..79c96811863f (1 commits) (flutter/engine#6682)


The AutoRoll server is located here: https://autoroll.skia.org/r/flutter-engine-flutter-autoroll

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/&#43;/master/autoroll/README.md

If the roll is causing failures, please contact the current sheriff, who should
be CC&#39;d on the roll, and stop the roller if necessary.

dnfield committed:
Revert to last known good engine (#23700)

One of the devicelab tests is broken after this commit - rolling to last known good commit while I investigate.

/cc @bkonyi

awdavies committed:
[frdp] Add `includeNonFlutterIsolates` flag. (#23693)

Previously there was no way to add this flag when searching for
Isolates.

jacob314 committed:
Add option to track widget rebuilds and repaints from the Flutter inspector. (#23534)

NurlashKO committed:
Add examples to documentation of TimeOfDay class. (#23452)

bkonyi committed:
Added '--check-for-remote-artifacts' option for Flutter Doctor. (#23543)

* Added '--check-for-remote-artifacts' option for Flutter Doctor.

This option takes a Flutter engine revision and issues HEAD requests to
determine whether or not artifacts for the provided engine revision are
available from cloud storage. This functionality is needed for the Dart
SDK autoroller to avoid creating a PR to roll the engine into the
framework while artifacts haven't finished building, which would cause
Cirrus tests to fail.

Commit:
Updated supportedLocales doc, etc (#23683)

devoncarew committed:
remove unnecessary overrides (#23650)

aam committed:
Reland change that speeds up multiple devices hot-reload (#23695)

* Revert "Revert "Run reload asynchronously so that multiple devices can reload in parallel. (#22693)" (#23598)"

This reverts commit 0b68068d6a33f35cec568ae6a9b394ff2a9fdb76.

* Fix refreshViews so it sends app-wide(rather than per-isolate) service request.

Sending per-isolate request caused dead-lock in the engine in case of more-than-one ui isolate.

jslavitz commented:
@dudeofawesome Are you still working on this? Awesome work so far by the way!

Hixie committed:
[H] Cleanup (#23632)

* Avoid abbreviations

* Sample code for AppBar.leading

* Add a test for OverflowBox/FractionallySizedBox

* Minor wording improvements in the text.

The words "note that" here don't really contribute to the flow.

Hixie committed:
Allow ChangeNotifier to be mixed in again (#23631)

Luckily this class didn't actually need to extend its superclass, it
only implements the interface. So we can change `extends` to
`implements` and that's close enough, while allowing the class to be
mixed in again.

GaryQian committed:
Default locales test for flutter tester. Roll engine 162b2e98c03..7be0217d67969 (#23596)

162b2e98c036a8f8c355fd475dbe0db4f4bc886c..7be0217d6796954694753647b7a757b8e3907f64
7be0217d6 Roll back _countryCode assert - breaking change (#6693)
91a019cfd Fix popSystemNavigator (#6691)
2b2fbf0f6 Add Locale.fromSubtags and support for scriptCode. (#6518)
58c8e30e5 Roll src/third_party/skia ab18c8e6cc20..d48b7a881b24 (5 commits) (#6690)
3b17cfb68 Flutter tester default locales (#6689)
96bbd2b36 Roll buildroot to 11a934e99eaa4aa8e278cd2772aff4f51f1f3c41 (#6687)
0e9defbd6 Roll src/third_party/skia 68825776f4b4..ab18c8e6cc20 (11 commits) (#6688)
cc686d7ae Don't populate the external view embedder in PaintContext. (#6686)
ab782faa7 Roll src/third_party/skia 1de48d8040aa..68825776f4b4 (1 commits) (#6685)
c4aa8d343 Roll src/third_party/skia 797197a772b8..1de48d8040aa (2 commits) (#6684)
7352251eb Roll src/third_party/skia 79c96811863f..797197a772b8 (1 commits) (#6683)
3dac47e4d Roll src/third_party/skia 38e4fd0c5654..79c96811863f (1 commits) (#6682)
2f06a53a8 Roll src/third_party/skia b53f1f46982d..38e4fd0c5654 (1 commits) (#6681)
a1d7cad70 Fix inconsistent include syntax (#6680)
236661cd8 Roll src/third_party/skia 3b79aa3a5ad0..b53f1f46982d (13 commits) (#6679)
c4f50611d Roll buildroot to pick up Mac toolchain updates. (#6678)
ba8f6aa71 Handle Windows headers defining ERROR to 0 in log levels. (#6677)
505d2a923 Roll buildroot to pick up updates to custom toolchains. (#6674)
6c2a0b3a6 Undefine ERROR in platform_view_layer.cc (#6675)
55e12993a Update FlutterPlugin.h docs, suppress warning for older API (#6672)
f7970048d Attach and position embedded UIVIews (#6614)
df85722fa Plumb the iOS PlatformViewsController into flow. (#6603)
2bfb893cf iOS Embedding Refactor (#6447)

devoncarew committed:
add explicit const keywords (#23652)

engine-flutter-autoroll committed:
Roll engine 7be0217d6796..3a6775730028 (3 commits) (#23726)


https://github.com/flutter/engine/compare/7be0217d6796...3a6775730028

git log 7be0217d6796..3a6775730028 --no-merges --oneline
3a6775730 Roll src/third_party/skia 14768f64fba3..dde307f14273 (3 commits) (flutter/engine#6697)
ab92772e8 Roll src/third_party/skia ca44d5d85895..14768f64fba3 (1 commits) (flutter/engine#6695)
d7e2be279 Roll src/third_party/skia d48b7a881b24..ca44d5d85895 (3 commits) (flutter/engine#6694)


The AutoRoll server is located here: https://autoroll.skia.org/r/flutter-engine-flutter-autoroll

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/&#43;/master/autoroll/README.md

If the roll is causing failures, please contact the current sheriff, who should
be CC&#39;d on the roll, and stop the roller if necessary.

mit-mit committed:
Update template comments (#23665)

* Update template comments

Make template comments consistent with recent tools changes

* Update packages/flutter_tools/templates/app/lib/main.dart.tmpl

Co-Authored-By: mit-mit <mit-mit@users.noreply.github.com>

* Fix merge conflict

dnfield committed:
Reland Xcode backend refactor (#23574)

* Use Xcode build configurations to drive Flutter build mode

* Proper check wrt local_engine, print error if profile mode misisng

* Remove unused code, update tests, fix template problem, update warning

* fix up warning

* add explanatory dev comment

* fix whitespace

* missing words, change lambda arrow to function body

* error indentation

* Test early exits for xcode_backend.sh

* only on macOS, use right test

* Update error messages

* case insensitive compare for build config

* Update gallery podfile

* update projects to add profile configuration

* make compatible with flavors

* add missing plist files

Commit:
Verify that date/number translations exist for all supported languages (#23692)

Hixie committed:
Update templates (#23698)

Removes some `new`s and other minor cleanup.

dnfield committed:
Revert "Reland Xcode backend refactor" (#23737)

* Revert "Update templates (#23698)"

This reverts commit 5b0de6d679ff2e870711b986577bc7769c1745fb.

* Revert "Verify that date/number translations exist for all supported languages (#23692)"

This reverts commit 3449edf256aa389405ef6a86387dc02c81061b2e.

* Revert "Reland Xcode backend refactor (#23574)"

This reverts commit 7270f2845d9a975387191ab6adcada4c492273b5.

Commit:
Resolve MaterialApp.locale, WidgetsApp.locale per supportedLocales (#23707)

DanTup committed:
Add a hot reload test that modifies code and verifies it executes (#23725)

Commit:
More debugCheckHasMaterialLocalizations() asserts (#23740)

jslavitz committed:
Adds Text Direction functionality to the TextField (#23609)

* added textDirection to TextField

GaryQian committed:
Remove '' as countryCode in localization tests (#23709)

devoncarew committed:
update the docs for the module template (#23699)

* update the docs for the module template

* update text

jslavitz committed:
Adds sample code to TextFormField (#23680)

* Adds sample code to TextFormField

gspencergoog committed:
Stop wrapping adb, gradle and ios logger output, and update terminal wrapping column dynamically. (#23592)

Subcommand output (gradle, adb, etc) is no longer wrapped, and wrapping notices when the terminal column width changes dynamically now.

Fixes #23267.
Fixes #23266.

gspencergoog committed:
Allow users to create samples using flutter create. (#23584)

This adds flutter create --sample which allows users to execute a command which will create a working sample app from samples embedded in the API docs.

The command looks something like this:

flutter create --sample=chip.DeletableChipAttributes.onDeleted mysample

johnsonmh committed:
[Material] Create TabBarTheme.of constructor (#23702)

* Create TabBarTheme of(...) constructor

* First round comments

* fix imports

dudeofawesome commented:
I'd love to be able to finish this PR up, but I've been unable to get my dev environment set up properly. 

I'm trying to use the Flutter executable in this repository, which I've forked and then cloned on to my machine at `~/Github/flutter`, with the executable at `~/Github/flutter/bin/flutter`. If I run `$ ~/Github/flutter/bin/flutter --version` I get the following: 
```
Flutter 0.0.21-pre.331 • channel master • https://github.com/dudeofawesome/flutter.git
Framework • revision a3074304b9 (4 days ago) • 2018-10-26 22:39:08 -0700
Engine • revision 6c2a0b3a6d
Tools • Dart 2.1.0 (build 2.1.0-dev.8.0 bf26f760b1)
```
which is clearly a problem given Flutter is at `0.10.1` or something now. 

I'm not really sure how to get the bundled version of Flutter updated. I've tried merging in the most recent changes from `upstream/master`, but that didn't change the version that `$ ~/Github/flutter/bin/flutter --version` reports.

jslavitz commented:
@dudeofawesome Awesome! And I've had that exact problem. I've found the best way to resolve it is to get the diff file for the current PR and then use git apply to merge it into the current version of flutter. 

To get the diff file (if you don't already know), you have to click the "files changed" link at the top of the pr, delete the "/files" and replace it with ".diff". Press enter and you'll get the diff file! (Also, you'll probably need to add a new line to the end of the file you make.) If you then copy that text into a text file, you can use `diff apply <file>` to patch it into a new branch. 

If git apply fails, you'll have to manually copy your changes into a new version of flutter. 

Let me know if that helps!

dudeofawesome committed:
Merge branch 'master' of https://github.com/dudeofawesome/flutter

dudeofawesome committed:
🐛 fix rebase issues

dudeofawesome committed:
Merge branch 'feat/user-account-arrow-anim' of https://github.com/dudeofawesome/flutter into feat/user-account-arrow-anim

dudeofawesome commented:
@jslavitz I ended up trying a rebase (which I've tried before) and it worked this time! I've now got flutter `0.10.2-pre.131`!

dudeofawesome committed:
🎨  remove unnecessary new keywords

jslavitz commented:
@dudeofawesome Great! Though, adding those files isn't good and we won't be able to merge this commit unless it only has the changes from the files you were working on :/. It would probably make the most sense to strip out your changed file and then merge that into a fresh master. (You also might have to make a new PR because you unfortunately can't un-commit files from a PR :( ).

jslavitz commented:
@dudeofawesome If you're busy right now, I can take this one over! Just wondering if you modified any other files than `user_accounts_drawer_header.dart`?

dudeofawesome commented:
@jslavitz I'm a little bit stuck as far as trying to figure out exactly how I should go about writing tests for this. The only user-facing change here is that the arrow now animates. Is there any prior art that you know of as far as testing animations?

I've only modified the `user_accounts_drawer_header.dart` file.

jslavitz commented:
@dudeofawesome So the purpose of tests are just to verify that the code you've written works the way you expect it to. So in this case, you probably want to test that that:
 - Before the header drawer is tapped, the arrow is fully down
 - In the middle of the animation, the arrow is neither up or down
 - At the end of the animation, the arrow is fully up.
And then have these three tests in reverse direction for when the arrow goes from up to down. You'll want to add the tests to the `user_accounts_drawer_header_test.dart` file. For examples on how to write this test, you'll need to look in other test files. To find those files, you could `grep` for `AnimationController` in the framework to find files that have animations in them. Then you could find the test files relating to the files and see how the animations are tested there.

If you want help, feel free to respond to comment or if you want me to take over let me know! Good luck!
